app/code/Magento/CatalogUrlRewrite/Model/ProductScopeRewriteGenerator.php

```diff
@@ -276,4 +283,83 @@ private function isCategoryRewritesEnabled()
     {
         return (bool)$this->config->getValue('catalog/seo/generate_category_product_rewrites');
     }
+
+    /**
+     * Return product changed categories store root category ids
+     *
+     * @param Product $product
+     * @return int[]
+     */
+    private function getChangedCategoriesStoreIds(Product $product): array
+    {
+        $oldCategoryIds = $product->getOrigData('category_ids');
+        $actualCategoryIds = $product->getCategoryIds();
+        $result = [];
+        if ($oldCategoryIds === null) {
+            return $result;
+        }
+        if (array_diff($actualCategoryIds, $oldCategoryIds)) {
+            $result = $this->getAddedCategoriesStoreIds($product);
+        }
+        if (array_diff($oldCategoryIds, $actualCategoryIds)) {
+            $result = array_merge($result, $this->getRemovedCategoriesStoreIds($product));
+        }
+
+        return array_map('intval', array_unique($result));
+    }
+
+    /**
+     * Return added product categories store root category ids
+     *
+     * @param Product $product
+     * @return array
+     */
+    private function getAddedCategoriesStoreIds(Product $product): array
+    {
+        $newCategories = array_filter(
+            $product->getCategoryCollection()->getItems(),
+            function (CategoryInterface $category) use ($product) {
+                return !in_array($category->getId(), $product->getOrigData('category_ids'));
+            }
+        );
```

Comment:
Is it possible to use here collection filter instead of post-processing?
```suggestion
        $newCategories = $product->getCategoryCollection()->addIdFilter($product->getOrigData('category_ids'))->getItems();
```

Reply:
it is not, as in that case, we can get either the wrong result if the collection was loaded before or break the original collection for further processing. Reload items here does not make sense.